### PR TITLE
add nvm counters

### PR DESCRIPTION
**Pull request by jpbland1**

### CODE REVIEW

File: src/wh_server_counter.c [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    switch (action)
+    {
+    case WH_COUNTER_RESET:
+        /* write 0 to nvm with the supplied id and user_id */
+        counter = 0;
+        memset((void*)meta, 0, sizeof(meta));
+        meta->id = MAKE_WOLFHSM_KEYID(WOLFHSM_KEYTYPE_COUNTER,
+            server->comm->client_id, packet->counterResetReq.keyId);
+        meta->len = sizeof(counter);
+        ret = wh_Nvm_AddObject(server->nvm, meta, meta->len,
```
Comment (billphipps) [inline src/wh_server_counter.c:49]: I thought we were going to store the counter value in the label so we didn't need any data segment?

File: src/wh_server_counter.c [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        if (ret == 0) {
+            ret = wh_Nvm_Read(server->nvm, meta->id, 0, sizeof(counter),
+                (uint8_t*)&counter);
+        }
+        /* increment and write the counter back */
+        if (ret == 0) {
+            counter++;
+            /* check if counter rolled over */
+            if (counter == 0)
+                ret = WH_ERROR_NOSPACE;
```
Comment (billphipps) [inline src/wh_server_counter.c:71]: This error is incorrect and conflicts with the real possibility of not having any free entries in the directory (which the user should immediately attempt to reclaim space using DestroyObjects).  The counter should be defined as saturating so if it is going to roll over, it simply stays at the highest value and does not return an error.  Since you return the value, the user can check

File: src/wh_server_counter.c [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        }
+        break;
+    case WH_COUNTER_INCREMENT:
+        /* read the counter's metadata */
+        ret = wh_Nvm_GetMetadata(server->nvm, MAKE_WOLFHSM_KEYID(
+            WOLFHSM_KEYTYPE_COUNTER, server->comm->client_id,
+            packet->counterIncrementReq.keyId), meta);
+        /* read the counter value */
+        if (ret == 0) {
+            ret = wh_Nvm_Read(server->nvm, meta->id, 0, sizeof(counter),
```
Comment (billphipps) [inline src/wh_server_counter.c:63]: If the counter is left in the metadata label, then you won't have to Nvm_Read here.  Just get the metadata and decode the label

File: src/wh_server_counter.c [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    whPacket* packet = (whPacket*)data;
+    whNvmMetadata meta[1] = {0};
+
+    if (server == NULL || server->nvm == NULL || data == NULL || size == NULL)
+        return WH_ERROR_BADARGS;
+
+    switch (action)
+    {
+    case WH_COUNTER_RESET:
+        /* write 0 to nvm with the supplied id and user_id */
```
Comment (billphipps) [inline src/wh_server_counter.c:43]: There's no way to remove a counter from NVM.  Recommend to add a function to destroy a counter.

File: src/wh_server_counter.c [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        counter = 0;
```
Comment (billphipps) [inline src/wh_server_counter.c:44]: Consider letting the user specify the starting value.

File: test/wh_test_crypto.c [hunk 6]
```diff
@@ -562,6 +561,32 @@ int whTest_CryptoClientConfig(whClientConfig* config)
...
         goto exit;
     }
+    printf("CMAC SUCCESS\n");
+    /* test counters */
+    keyId = 1;
+    if ((ret = wh_Client_CounterReset(client, keyId, &counter)) != 0 || counter != 0) {
+        WH_ERROR_PRINT("Failed to wh_Client_CounterReset %d\n", ret);
+        goto exit;
+    }
+    if ((ret = wh_Client_CounterIncrement(client, keyId, &counter)) != 0 || counter != 1) {
```
Comment (billphipps) [inline test/wh_test_crypto.c:571]: Make sure to add a loop that invokes Increment more times than entries in the directory (currently 32), so this call will return with WH_ERROR_NOSPACE.  You should add a wh_Client_NvmDestroyObjects(client, 0,NULL,&out_rc) to reclaim the space and retry the CounterIncrement.  Just noticed the DestroyObjects API is incorrect.

**Review by billphipps — CHANGES_REQUESTED** [review]
Looks good.  a few changes:
1. Please move the counter value into the metadata label. 
2. Consider automatically handling the reclaimable space error when incrementing.

File: wolfhsm/wh_packet.h [hunk 7]
```diff
@@ -307,6 +307,36 @@ typedef struct WOLFHSM_PACK wh_Packet_version_exchange
...
+typedef struct WOLFHSM_PACK wh_Packet_counter_read_req
+{
+    uint16_t keyId;
+} wh_Packet_counter_read_req;
+
+typedef struct WOLFHSM_PACK wh_Packet_counter_read_res
+{
+    uint32_t counter;
+} wh_Packet_counter_read_res;
+
```
Comment (bigbrett) [inline wolfhsm/wh_packet.h:339]: hmm, I'm noticing we have a split paradigm developing here - you have been putting messages into the `whPacket` union in this file, like you do for crypto, whereas Bill and I have been adding new message definitions and the necessary marshalling/data transformations in their own `wh_message_xxx.{c,h}` files. I think we should standardize on doing it one way. 

Related, I see there aren't any endian transformations for these messages. This will break if we ever have platforms with mixed HSM and host endianness. The reason Bill and I implemented the intermediate message layer is to allow a distinct module that handles the endian transformations for a given message data type, between host format and HSM format. 

I'm also now realizing we don't do that for any of the crypto...I think that is something we will need to add at some point.

I suppose we could punt it for now, though I'd really prefer adding the appropriate transformation support to this module before merging, and just punting the crypto. And separating this from crypto in its entirety, as it is a completely different feature/module

**Review by bigbrett — CHANGES_REQUESTED** [review]
Nice. A few comments and things to address. Also, I would really really like to see tests do a little more exercising of the functionality, testing edge cases, etc. rather than just incrementing a counter once and resetting it. Doing some incrementation in a loop, demonstrating proper handling of rollover/saturation, etc would be really good. Tests that could catch regressions to the logic are awesome.

File: wolfhsm/wh_packet.h (hunk 7 above)
Comment (jpbland1) [reply wolfhsm/wh_packet.h]: I think everything should be put in packet.h and all packets should be defined in the single unionionized struct that I've been using so we don't have to serialize and deserialize anything and can use a single piece of memory for the packet by just casting. We will need to add networkize and hostize functions that swap the endianness in a single switch statement based on message group and action. frankly adding new files for all this just makes more technical debt

**Comment by jpbland1** [comment]
Updated based on you guys feedback, I had to modify nvm to allow NULL data entries for our metadata only idea, need to confirm this doesn't have any unintended consequences

File: src/wh_client.c [hunk 8]
```diff
@@ -733,6 +733,203 @@ int wh_Client_KeyErase(whClientContext* c, whNvmId keyId)
...
+    if (ret == 0) {
+        if (packet->rc != 0)
+            ret = packet->rc;
+        else
+            *counter = packet->counterInitRes.counter;
+    }
+    return ret;
+}
+
+int wh_Client_CounterInit(whClientContext* c, whNvmId keyId, uint32_t* counter)
```
Comment (billphipps) [inline src/wh_client.c:771]: Consider splitting the input counter and the output counter parameters and allow null for output counter.

File: src/wh_client.c [hunk 9]
```diff
@@ -733,6 +733,203 @@ int wh_Client_KeyErase(whClientContext* c, whNvmId keyId)
...
+}
+
+int wh_Client_CounterIncrementResponse(whClientContext* c, uint32_t* counter)
+{
+    uint16_t group;
+    uint16_t action;
+    uint16_t size;
+    int ret;
+    whPacket* packet;
+    if (c == NULL || counter == NULL)
```
Comment (billphipps) [inline src/wh_client.c:819]: Recommend to allow NULL output counter

File: src/wh_client.c [hunk 10]
```diff
@@ -733,6 +733,203 @@ int wh_Client_KeyErase(whClientContext* c, whNvmId keyId)
...
+        return WH_ERROR_BADARGS;
+    packet = (whPacket*)wh_CommClient_GetDataPtr(c->comm);
+    /* set keyId */
+    packet->counterDestroyReq.keyId = keyId;
+    return wh_Client_SendRequest(c, WH_MESSAGE_GROUP_COUNTER,
+        WH_COUNTER_DESTROY, WOLFHSM_PACKET_STUB_SIZE +
+        sizeof(packet->counterReadReq), (uint8_t*)packet);
+}
+
+int wh_Client_CounterDestroyResponse(whClientContext* c)
```
Comment (billphipps) [inline src/wh_client.c:903]: If there could be ambiguity in the return value between local and remote, then you should provide the local return value as the return to this function and use and output parameter for the remote return value.

File: src/wh_server_counter.c [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    switch (action)
+    {
+    case WH_COUNTER_INIT:
+        /* write 0 to nvm with the supplied id and user_id */
+        memset((void*)meta, 0, sizeof(meta));
+        meta->id = MAKE_WOLFHSM_KEYID(WOLFHSM_KEYTYPE_COUNTER,
+            server->comm->client_id, packet->counterInitReq.keyId);
+        /* use the label buffer to hold the counter value */
+        *(uint32_t*)meta->label = packet->counterInitReq.counter;
+        ret = wh_Nvm_AddObject(server->nvm, meta, 0, NULL);
```
Comment (billphipps) [inline src/wh_server_counter.c:49]: Shouldn't Init return an error if the object already exists?  Otherwise, this simply overwrites the current value.

File: src/wh_server_counter.c [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        /* if we ran out of space try cleaning up duplicate entries and then retry */
```
Comment (billphipps) [inline src/wh_server_counter.c:50]:
```suggestion
        /* if we ran out of space try reclaiming space and then retry */
```

File: src/wh_server_counter.c [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+        ret = wh_Nvm_GetMetadata(server->nvm, MAKE_WOLFHSM_KEYID(
+            WOLFHSM_KEYTYPE_COUNTER, server->comm->client_id,
+            packet->counterIncrementReq.keyId), meta);
+        /* increment and write the counter back */
+        if (ret == 0) {
+            (*(uint32_t*)meta->label)++;
+            /* set counter to uint32_t max if it rolled over */
+            if (*(uint32_t*)meta->label == 0)
+                *(uint32_t*)meta->label = 0xffffffff;
+            ret = wh_Nvm_AddObject(server->nvm, meta, 0, NULL);
```
Comment (billphipps) [inline src/wh_server_counter.c:72]: No need to AddObject if the count is saturated.

File: src/wh_server_counter.c [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+        /* write 0 to nvm with the supplied id and user_id */
+        memset((void*)meta, 0, sizeof(meta));
+        meta->id = MAKE_WOLFHSM_KEYID(WOLFHSM_KEYTYPE_COUNTER,
+            server->comm->client_id, packet->counterInitReq.keyId);
+        /* use the label buffer to hold the counter value */
+        *(uint32_t*)meta->label = packet->counterInitReq.counter;
+        ret = wh_Nvm_AddObject(server->nvm, meta, 0, NULL);
+        /* if we ran out of space try cleaning up duplicate entries and then retry */
+        if (ret == WH_ERROR_NOSPACE) {
+            ret = wh_Nvm_DestroyObjects(server->nvm, 0, NULL);
```
Comment (billphipps) [inline src/wh_server_counter.c:52]: Should check to see if there is any space that could be reclaimed GetAvailable() before invoking the painful DestroyObjects().

File: src/wh_server_counter.c [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+            WOLFHSM_KEYTYPE_COUNTER, server->comm->client_id,
+            packet->counterReadReq.keyId), meta);
+        /* return counter to the caller */
+        if (ret == 0) {
+            packet->counterReadRes.counter = *(uint32_t*)meta->label;
+            *size = WOLFHSM_PACKET_STUB_SIZE + sizeof(packet->counterReadRes);
+        }
+        break;
+    case WH_COUNTER_DESTROY:
+        keyId = packet->counterDestroyReq.keyId;
```
Comment (billphipps) [inline src/wh_server_counter.c:99]: Should this be MAKE_WOLFHSM_KEYID(WOLFHSM_KEYTYPE_COUNTER, ... ) ?

File: wolfhsm/wh_client.h [hunk 16]
```diff
@@ -617,6 +617,82 @@ int wh_Client_AesCmacVerify(Cmac* cmac, const byte* check, word32 checkSz,
...
+    uint32_t counter);
+int wh_Client_CounterInitResponse(whClientContext* c, uint32_t* counter);
+/**
+ * @brief Creates and initializes a counter with the value set in counter.
+ *
+ * This function creates/resets a counter with the supplied keyId and gives it
+ * the value stored in counter at the start of the call.
+ *
+ * @param[in] c Pointer to the whClientContext structure.
+ * @param[in] keyId Key ID to be associated with the CMAC key.
```
Comment (billphipps) [inline wolfhsm/wh_client.h:631]: Copy pasta with the CMAC.  Consider naming the parameter CounterId instead of KeyId.

**Review by billphipps — CHANGES_REQUESTED** [review]
Maybe a bug in the destroy function?  Also, the overall intent of these counters was to provide a basis to implement the PKCS11 monotonic counters (https://docs.oasis-open.org/pkcs11/pkcs11-base/v3.0/csprd01/pkcs11-base-v3.0-csprd01.html#_Toc10472588), so let's make sure that we can simulate the same behavior using the exposed API and internals.

File: wolfhsm/wh_packet.h (hunk 7 above)
Comment (billphipps) [reply wolfhsm/wh_packet.h]: Counters should be supported even when crypto is not.  We'll need to move things around in the future, but we'll stay the course for now.  Providing separate translation units means fewer conditional compilation flags, so the giant union isn't always the easiest way to manage bulk data.  Each individual kind of message will be typecast anyway, and should be processed in separate functions, so the union only exists at the main dispatch.

Again, nothing to fix at this point, but when we get a chance, we really should focus on eliminating "packed" by reorganizing the structs and using fixed size types in message fields.  Then, the only processing per field is for endianness.

Comment (bigbrett) [reply wolfhsm/wh_packet.h]: Well I think we will always need to serialize/deserialize. To clarify, when I say serialize/deserialize I'm referring to the same thing as you are when you say networkize/hostize. I suppose "marshalling" is a better word to use. Going from a standard "wire format" to a native host format. So if we want to support platforms with different byte orders, there isn't really any way we can get around this operation. And it goes beyond endianness too, it includes packing/padding of structs, which can be annoyingly different for different compilers/toolchains/architectures.

And the marshaling is already is done based on the message group and action like you are suggesting, but it just is handled one level down (e.g. NVM messages are deserialized in `wh_Server_HandleNvmRequest`, Comm messages are deserialized in `wh_Server_HandleCommRequest`, etc.) instead of before the message dispatch in `wh_Server_HandleRequest()`. Not sure why doing it this way results in more technical debt. Some might say more files mean less code in a given context, making things clearer, and there is more of a clear separation of concerns. Not at all trying to die on this hill, I just don't think it is necessarily a problem. What is a problem IMO, is not doing any proper marshalling of data and manually reordering/padding the structs to try and avoid issues.

That said, I've long been thinking about using a ser/de library (or making our own simple one) so wolfHSM itself doesn't need to worry about this stuff. Something like protobuf/flatbuffers or the like. Since using packed structs on different architectures is pretty dangerous, and right now we are just manually packing/aligning structs which seems like not the best way.

Ideally we have a struct representation of each data type that is a normal native struct able to be used on both sides, and we can just call `myStruct_deSerialize(&myLocalStruct, &commBuffer)` or `myStruct_serialize(&myLocalStruct, &commBuffer)`  and the data would be properly marshalled without worrying about endianness, padding, alignment, etc.

Add it to the backlog, I suppose...

**Comment by jpbland1** [comment]
rebased, need to discuss the issue with `wh_Nvm_DestroyObjects` the way I'm using it

File: src/wh_nvm.c [hunk 17]
```diff
@@ -87,6 +87,36 @@ int wh_Nvm_GetAvailable(whNvmContext* context,
             out_reclaim_size, out_reclaim_objects);
 }
 
+int wh_Nvm_AddObjectPessimistic(whNvmContext* context, whNvmMetadata *meta,
```
Comment (billphipps) [inline src/wh_nvm.c:90]: Let's try a more objective name.  How about:
```suggestion
int wh_Nvm_AddObjectWithReclaim(whNvmContext* context, whNvmMetadata *meta,
```

File: src/wh_nvm.c [hunk 18]
```diff
@@ -87,6 +87,36 @@ int wh_Nvm_GetAvailable(whNvmContext* context,
...
+    const uint8_t* data, whNvmSize dataLen)
```
Comment (billphipps) [inline src/wh_nvm.c:91]: Keep the same argument order
```suggestion
    whNvmSize dataLen, const uint8_t* data)
```

File: src/wh_nvm.c [hunk 19]
```diff
@@ -87,6 +87,36 @@ int wh_Nvm_GetAvailable(whNvmContext* context,
...
+    int ret;
+    uint32_t availableSize;
+    uint32_t reclaimSize;
+    if (context == NULL)
+        return WH_ERROR_BADARGS;
+    /* check if we have room for the object */
+    ret = wh_Nvm_GetAvailable(context, &availableSize, &availableObjects,
+        &reclaimSize, &reclaimObjects);
+    if (ret == 0) {
+        if (availableSize >= WOLFHSM_NVM_METADATA_LEN + dataLen &&
```
Comment (billphipps) [inline src/wh_nvm.c:104]: Metadata len is not included in the available data space
```suggestion
        if (availableSize >=  dataLen &&
```

File: src/wh_nvm.c [hunk 20]
```diff
@@ -87,6 +87,36 @@ int wh_Nvm_GetAvailable(whNvmContext* context,
...
+            availableObjects > 0) {
+            return wh_Nvm_AddObject(context, meta, dataLen, data);
```
Comment (billphipps) [inline src/wh_nvm.c:106]: Prefer single return
```suggestion
            ret = wh_Nvm_AddObject(context, meta, dataLen, data);
```

File: src/wh_nvm.c [hunk 21]
```diff
@@ -87,6 +87,36 @@ int wh_Nvm_GetAvailable(whNvmContext* context,
...
+        }
+        /* if there's no room, try to reclaim space, otherwise return error */
+        else if (availableSize + reclaimSize >= WOLFHSM_NVM_METADATA_LEN +
+            dataLen && reclaimObjects > 0) {
+            ret = wh_Nvm_DestroyObjects(context, 0, NULL);
+            if (ret == 0)
+                return wh_Nvm_AddObject(context, meta, dataLen, data);
```
Comment (billphipps) [inline src/wh_nvm.c:113]:
```suggestion
                ret = wh_Nvm_AddObject(context, meta, dataLen, data);
```

File: src/wh_nvm.c [hunk 22]
```diff
@@ -87,6 +87,36 @@ int wh_Nvm_GetAvailable(whNvmContext* context,
...
+    ret = wh_Nvm_GetAvailable(context, &availableSize, &availableObjects,
+        &reclaimSize, &reclaimObjects);
+    if (ret == 0) {
+        if (availableSize >= WOLFHSM_NVM_METADATA_LEN + dataLen &&
+            availableObjects > 0) {
+            return wh_Nvm_AddObject(context, meta, dataLen, data);
+        }
+        /* if there's no room, try to reclaim space, otherwise return error */
+        else if (availableSize + reclaimSize >= WOLFHSM_NVM_METADATA_LEN +
+            dataLen && reclaimObjects > 0) {
```
Comment (billphipps) [inline src/wh_nvm.c:110]:
```suggestion
        else if (availableSize + reclaimSize >= dataLen && reclaimObjects > 0) {
```

File: src/wh_server_counter.c [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+    whNvmMetadata meta[1] = {0};
+
+    if (server == NULL || server->nvm == NULL || data == NULL || size == NULL)
+        return WH_ERROR_BADARGS;
+
+    switch (action)
+    {
+    case WH_COUNTER_INIT:
+        /* write 0 to nvm with the supplied id and user_id */
+        memset((void*)meta, 0, sizeof(meta));
```
Comment (billphipps) [inline src/wh_server_counter.c:44]: Already zeroed above.
```suggestion
```

File: src/wh_server_counter.c [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+#include "wolfhsm/wh_server_counter.h"
+#include "wolfhsm/wh_message.h"
+#include "wolfhsm/wh_packet.h"
+#include "wolfhsm/wh_error.h"
+
+int wh_Server_HandleCounter(whServerContext* server, uint16_t action,
+    uint8_t* data, uint16_t* size)
+{
+    whKeyId counterId;
+    int ret;
```
Comment (billphipps) [inline src/wh_server_counter.c:33]:
```suggestion
    whKeyId counterId = 0;
    int ret = 0;
```

File: src/wh_server_keystore.c [hunk 25]
```diff
@@ -304,8 +304,16 @@ int hsmCommitKey(whServerContext* server, whNvmId keyId)
...
+    ret = wh_Nvm_AddObjectPessimistic(server->nvm, cacheSlot->meta,
+        cacheSlot->buffer, cacheSlot->meta->len);
+    /* if we ran out of space clean out duplicate entries and retry */
+    if (ret == WH_ERROR_NOSPACE) {
+        ret = wh_Nvm_DestroyObjects(server->nvm, 0, NULL);
+        if (ret == 0) {
+            ret = wh_Nvm_AddObject(server->nvm, cacheSlot->meta,
+                cacheSlot->meta->len, cacheSlot->buffer);
+        }
+    }
```
Comment (billphipps) [inline src/wh_server_keystore.c:316]: Duplicate functionality.  Remove.
```suggestion
```

**Review by billphipps — CHANGES_REQUESTED** [review]
A couple of tweaks necessary but looks good.

File: src/wh_nvm.c (hunk 18 above)
Comment (jpbland1) [reply src/wh_nvm.c:91]: I'll change it for now but I think all the functions should use buffer, length (like all the woflCrypt, wolfssl, wolfBoot etc. functions)

**Review by billphipps — APPROVED** [review]

**Review by bigbrett — APPROVED** [review]
